File: programmers/Level 2/h-index.js
/*
programmers Level 2: H-index
2021-12-26
*/
function solution(citations) {
    let answer = 0;
    let num = citations.length;

    citations.sort((a, b) => {
        return a - b
    });

    for (let i = num; i >= 0; i--) {
        let index = citations.findIndex(v => v >= i);
        index = index === -1 ? num : index;

        if (num - index >= i) {
            answer = i
            break;
        }
    }

    return answer;
}

File: programmers/Level 2/h-index.cpp
/*
programmers Level 2: H-index
2021-12-26
*/
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

int solution(vector<int> citations) {
    int answer = 0;
    int num = citations.size();
    
    sort(citations.begin(), citations.end()); 
    
    for(int i = num ; i >= 0 ; i--) {
        int index = lower_bound(citations.begin(), citations.end(), i) - citations.begin();

        if(i <= num - index) {
            answer = i;
            break;
        }
    }
    
    return answer;
}

int main(){
    int answer = solution({3, 0, 6, 1, 5});
    cout << answer;

    return 0;
}